edb/schema/inheriting.py

```diff
@@ -378,12 +379,29 @@ def _classbases_from_ast(
 
         base_refs: List[so.InheritingObjectT] = []
         for b in getattr(astnode, 'bases', None) or []:
-            obj = utils.ast_to_object(
-                b,
-                modaliases=modaliases,
-                schema=schema,
-                metaclass=cls.get_schema_metaclass(),
-            )
+            try:
+                obj = utils.ast_to_object(
+                    b,
+                    modaliases=modaliases,
+                    schema=schema,
+                    metaclass=cls.get_schema_metaclass(),
+                )
+            except errors.InvalidReferenceError as initial_error:
+                from edb.schema import scalars as s_scalars
+                try:
+                    shell = utils.ast_to_object_shell(
+                        b,
+                        modaliases=modaliases,
+                        schema=schema,
+                        metaclass=s_scalars.ScalarType,
```

Comment:
Use `ast_to_object`, omit `metaclass` argument and use `get_verbosename()` on the returned object.  That way you don't hardcode the check for scalars and are able to check for general metaclass mismatch.

Reply:
1. I don't know what to do later. I can suggest `CREATE SCALAR TYPE` only for scalars. What `get_verbosename()` gives me is something like `scalar type 'xxx'`. Should I use `startswith`? (this doesn't work for enums) Should I use `get_schema_class_displayname()`?
2. Why use `ast_to_object` if that's a wrapper around shell with an extra step?

Reply:
If you want just the `SCALAR TYPE` part, then `get_schema_class_displayname()` would do.

Reply:
Done

Reply:
You still have the check hardcoded to scalars.  Remove `metaclass=` to make this work for any mismatched inheritance.

Reply:
Sorry, forgot this one. It doesn't change anything, though, because the hint is specifically for scalar types. I don't know what hint should be for other types. And basically I can't even recall what types can even be besides scalars and objects off the top of my head.

Reply:
All inheritance goes through this codepath, including links, constraints etc.

Reply:
So I'm not sure what else I should do? Any test that I can do for non-types?